### ISSUE TITLE
Deadlock / Race-condition after latest update

### ISSUE THREAD

**Issue by sehoffmann** [pytorch/data#1169]
### 🐛 Describe the bug

My datapipe that worked with 0.6 is currently hanging with the latest nightly version.

It's structured like this:
1. Main Thread Preprocess
2. Worker Process
3. Main Thread Postprocess

I use the MPRS with prefetch counts of 0 and do my own prefetching. My script can be found at https://github.com/sehoffmann/atmodata/blob/develop/examples/forecasting_benchmark.py and the relevant part that builds the datapipe can be found at https://github.com/sehoffmann/atmodata/blob/develop/atmodata/builder.py#L67.

What happens is that I receive 0-2 batches and then the pipe simply hangs. I was able to diagnose the location where the hanging occurs, namely the transfer from worker processes to the main thread (2. -> 3.).

To be more specifically:
(A) This sub-pipe: https://github.com/sehoffmann/atmodata/blob/develop/atmodata/builder.py#L105 works totally fine. I.e. each worker prefetches up to 10 batches
(B) This sub-pipe: https://github.com/sehoffmann/atmodata/blob/develop/atmodata/builder.py#L109 does not. I.e. for some reason it either yields none or only 1-2 batches and hangs.
(C) The post-processing pipe afterwards works fine. I.e. if (B) yields a batch it gets properly post-processed and forwarded to the final consumer.

Any suggestion on what might have caused this regression and what might help with further diagnosing are greatly appreciated.

Please note that the version in the repository monkey patched the worker_init_fn due to a limitation of the old version. In my local version, I deactivated that patch as it has since been become unnecessary. So this can't be the cause of the issue.

### Versions

nightly

**Comment by sehoffmann** [comment]
After further investigation, I am quite confident that this is a proper race-condition:
* If I only use a single worker process, the pipeline seems to be working fine
* With two worker processes, the pipeline yielded up to 93 batches before blocking/deadlocking.

This makes me believe, that we have a sequencing problem at hand whose chances of occurring increase with the number of worker processes.

**Comment by ejguan** [comment]
Thanks for flagging this issue. I might take a look at this problem later this week

**Comment by ejguan** [comment]
Can I know which nightly version are you using?

**Comment by sehoffmann** [comment]
Hey @ejguan 

I will try to reproduce the issue with a synthetic dataset later today. The original dataset is very big and also not easily available.

```
(wb) ➜  ~ python collect_env.py                                                                      
Collecting environment information...
PyTorch version: 2.1.0.dev20230427
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (Anaconda gcc) 11.2.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.17

Python version: 3.10.9 (main, Jan 11 2023, 15:21:40) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-3.10.0-1160.76.1.el7.x86_64-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 2080 Ti
Nvidia driver version: 515.65.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                72
On-line CPU(s) list:   0-71
Thread(s) per core:    2
Core(s) per socket:    18
Socket(s):             2
NUMA node(s):          2
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 85
Model name:            Intel(R) Xeon(R) Gold 6240 CPU @ 2.60GHz
Stepping:              7
CPU MHz:               3299.987
CPU max MHz:           3900.0000
CPU min MHz:           1000.0000
BogoMIPS:              5200.00
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              1024K
L3 cache:              25344K
NUMA node0 CPU(s):     0-17,36-53
NUMA node1 CPU(s):     18-35,54-71
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch epb cat_l3 cdp_l3 invpcid_single intel_ppin intel_pt ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req pku ospke avx512_vnni md_clear spec_ctrl intel_stibp flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==2.1.0.dev20230427
[pip3] torchaudio==2.1.0.dev20230427
[pip3] torchdata==0.7.0.dev20230427
[pip3] torchvision==0.16.0.dev20230427
[pip3] triton==2.1.0
[conda] blas                      1.0                         mkl    anaconda
[conda] cudatoolkit               11.8.0              h37601d7_11    conda-forge
[conda] libblas                   3.9.0            16_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            16_linux64_mkl    conda-forge
[conda] liblapack                 3.9.0            16_linux64_mkl    conda-forge
[conda] mkl                       2022.1.0           hc2b9512_224    anaconda
[conda] nomkl                     2.0                           0    anaconda
[conda] numpy                     1.23.5          py310h53a5b5f_0    conda-forge
[conda] pytorch                   2.1.0.dev20230427 py3.10_cuda11.8_cudnn8.7.0_0    pytorch-nightly
[conda] pytorch-cuda              11.8                 h7e8668a_5    pytorch-nightly
[conda] pytorch-mutex             1.0                        cuda    pytorch-nightly
[conda] torchaudio                2.1.0.dev20230427     py310_cu118    pytorch-nightly
[conda] torchdata                 0.7.0.dev20230427           py310    pytorch-nightly
[conda] torchtriton               2.1.0+7d1a95b046           py310    pytorch-nightly
[conda] torchvision               0.16.0.dev20230427     py310_cu118    pytorch-nightly
```

**Comment by sehoffmann** [comment]
@ejguan After further inspection, I now believe that the issue might stem from my custom serialization/deserialization code that transfers numpy arrays via shared memory. I will keep you up to date once I know more.

In case you want to test this yourself regardless: Checkout https://github.com/sehoffmann/atmodata/tree/bug/torchdata_mp and run `python -u examples/forecasting_benchmark.py --dataset ERA5 --path A --synthetic --workers 1`. With the synthetic dataset, the script somehow even fails to process the third shard. For your information: I'm processing 3 shards in parallel in a round robin fashion.

**Comment by sehoffmann** [comment]
State during deadlock:
* Worker process - Main thread:  eventloop.py:103
* Main process - Main thread: Prefetcher loop (prefetcher.py:96) at the very end of the graph
* Main process - QueueFeederThread: Condition.wait() from Queue._feed() in threading.py:320

Now something interesting: There's a PrefetcherIterDataPipe.thread_worker thread in the Worker process that has a `MemorySharer` source datapipe (custom datatype from me). But, this MemorySharer is the one that should get executed by the main proces before transfering it to the worker processes. Why is it in the worker process? In particular, it stems from this part:

```
 pipe = self.dataset

        if self.share_main_to_worker and self.n_workers > 0:
            pipe = pipe.share_memory()  # <--- this is the source datapipe of the thread_worker

        if self.dataloading_prefetch_cnt:
            pipe = pipe.prefetch(self.dataloading_prefetch_cnt)  # <--- this thread is in the worker process

        if self.caches_data:
            pipe = pipe.in_memory_cache()

        pipe = pipe.debug_print('main_pre', print_index=True, print_element=False)

        # 2. Worker process: Data processing
        if self.n_workers >= 1:
            if self.n_workers > 1:
                pipe = pipe.repeat(self.n_workers)
            pipe = pipe.sharding_round_robin_dispatch(SHARDING_PRIORITIES.MULTIPROCESSING)
            pipe = pipe.debug_print('worker_A', print_index=True, print_element=False)
 pipe = self.dataset

        if self.share_main_to_worker and self.n_workers > 0:
            pipe = pipe.share_memory()

        if self.dataloading_prefetch_cnt:
            pipe = pipe.prefetch(self.dataloading_prefetch_cnt)

        if self.caches_data:
            pipe = pipe.in_memory_cache()

        pipe = pipe.debug_print('main_pre', print_index=True, print_element=False)

        # 2. Worker process: Data processing
        if self.n_workers >= 1:
            if self.n_workers > 1:
                pipe = pipe.repeat(self.n_workers)
            pipe = pipe.sharding_round_robin_dispatch(SHARDING_PRIORITIES.MULTIPROCESSING)
            pipe = pipe.debug_print('worker_A', print_index=True, print_element=False)
```

**Comment by sehoffmann** [comment]
Hey @ejguan,

Indeed the issue with the synthetic dataset seems to be that my worker processes are killed by the OOM reaper and that this is not picked up by the main process. I opened a separate issue here: #1173 .

To clarify: This is different from the initial issue described here, and I will continue to try to come up with a reproducable example for you.

**Comment by sehoffmann** [comment]
For some unexplainable reason, I can no longer reproduce the issue, neither with real nor with synthetic data. However, there hasn't been any meaningful change I did that could have solved this.

I even rolled back to a previous commit and everything works fine now. I also increased the number of worker processes. This is even more surprising given that I experienced this issue on different compute nodes in a cluster environment. The only common denominator between those is the network FS that has caused issues in the past already. I am very bamboozled.

I will leave this issue open for a while in case the problem reoccurs; otherwise I will close it in a few days.

**Comment by sehoffmann** [comment]
I was able to confirm that hanging with very high likelihood occurred due to #1173
